docs/root/intro/version_history.rst

```diff
@@ -7,6 +7,9 @@ Version history
 * admin: added ability to configure listener :ref:`socket options <envoy_api_field_config.bootstrap.v2.Admin.socket_options>`.
 * admin: added config dump support for Secret Discovery Service :ref:`SecretConfigDump <envoy_api_msg_admin.v2alpha.SecretsConfigDump>`.
 * config: added access log :ref:`extension filter<envoy_api_field_config.filter.accesslog.v2.AccessLogFilter.extension_filter>`.
+* config: added support for :option:`--reject-unknown-fields-dynamic`, providing independent control
```

Comment:
The default is strict for everything today, right? IMO we should keep that as the default?

Reply:
I'm somewhat motivated here by the security angle. When we need to roll out new fields for security fixes, we want to be able to have control planes seamlessly push new configs. This is also what we want for the stable API versioning policy, where the same major version can have new unknown fields pushed without having to condition on client.

So, to support this, it makes sense to have most of the Envoys in the wild able to take unknown fields.

Given the whole original motivation for restricting unknown fields was dealing with hand-written static configs, I think we're in a pretty different place with control plane machine generated config.

Do any of these arguments sway you to relaxed by default for dynamic?

Reply:
> Do any of these arguments sway you to relaxed by default for dynamic?

TBH not really, especially since management servers may be doing a lot of potentially not great things to create "machine generated config." (Imagine they are reading in YAML and potentially dropping mistyped fields for example). For new users I think it's a substantially better experience if we fail by default on unknown fields.

With that said, if we can actually WARN on unknown fields, I would be interesting in switching my opinion. WDYT about defaulting to strict for now, and we can switch to permissive once we can WARN?

Reply:
I was planning on adding warn/stats in my next PR. The main issue with defaults is that we are using TClap switch args, which aren't really booleans, they are flags. So, to go from default strict to relaxed, we would have to rename the flags from "reject" to "accept" for the appropriate defaulting behavior. 

I could make this PR a bit larger and also do warning/stats here. WDYT?

Reply:
> I could make this PR a bit larger and also do warning/stats here.

+1. The other option would be to make the flag more of an enum that we could switch the default on? Similar to https://www.envoyproxy.io/docs/envoy/latest/operations/cli#cmdoption-mode? I'm fine either way. 

Or you could do the warning/stat PR first and then this one? Up to you.

Reply:
Yeah, let me do the warning/stat PR first. Will close out for now.

Reply:
Actually, scratch that. There are two things in this PR we will want anyway:
1. The fix for the regession in the strict/relax inversion.
2. All the test cases that were added. I rather have these around when doing warning/stats.
So, I'll just do a big PR.

Reply:
SGTM

/wait

Reply:
Warning log and stats added, PTAL.